src/main/java/com/aws/greengrass/deployment/DeploymentQueue.java

```diff
@@ -6,76 +6,162 @@
 package com.aws.greengrass.deployment;
 
 import com.aws.greengrass.deployment.model.Deployment;
-import com.aws.greengrass.deployment.model.Deployment.DeploymentType;
 import com.aws.greengrass.logging.api.Logger;
 import com.aws.greengrass.logging.impl.LogManager;
 
-import java.util.Iterator;
-import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
 
+/**
+ * <p>DeploymentQueue is a thread-safe deployment queue that automatically de-duplicates by deployment id, and
+ * also de-duplicates shadow deployments, i.e. there can be at-most-one shadow deployment enqueued.</p>
+ *
+ * <p>DeploymentQueue is implemented internally as a deployment id queue, along with a map of id -> deployment.</p>
+ *
+ * <p>When an offered deployment has a unique deployment id, it is enqueued normally.</p>
+ *
+ * <p>When an offered deployment has the same deployment id as an already-enqueued element:
+ *     - if the offered deployment meets replacement criteria, then the enqueued element is replaced by the offered
+ *       element, preserving queue order.
+ *     - otherwise, the offered deployment is ignored.</p>
+ *
+ * <p>Replacement criteria are as follows:
+ *     - offered.getDeploymentStage().getPriority() > enqueued.getDeploymentStage().getPriority(), OR
+ *     - offered.isCancelled() == true, OR
+ *     - offered.getDeploymentType().equals(SHADOW)</p>
+ *
+ * <p>When an offered deployment is a shadow deployment:
+ *     - if there is already a shadow deployment in the queue, then replace it and preserve queue order.
+ *     - otherwise, enqueue normally.</p>
+ */
 public class DeploymentQueue {
 
+    /**
+     * For the internal queue, shadow deployments use a special queue id. Effectively, this means there can be at most
+     * one shadow deployment in the queue.
+     */
+    private static final String SHADOW_DEPLOYMENT_QUEUE_ID = Deployment.DeploymentType.SHADOW.name();
+
+    private static final String DEPLOYMENT_ID_LOG_KEY = "DeploymentId";
+    private static final String DISCARDED_DEPLOYMENT_ID_LOG_KEY = "DiscardedDeploymentId";
     private static final Logger logger = LogManager.getLogger(DeploymentQueue.class);
-    private final ConcurrentLinkedQueue<Deployment> deploymentsQueue = new ConcurrentLinkedQueue();
 
     /**
-     * Add a deployment to the queue.
+     * Internal queue of deployment id's. For shadow deployments, the internal queue id is not the same as the
+     * actual deployment id.
+     */
+    private final Queue<String> deploymentIdQueue = new LinkedList<>();
+
+    /**
+     * Map of internal queue id -> deployment instance.
+     */
+    private final Map<String, Deployment> deploymentMap = new HashMap<>();
```

Comment:
could you just use a LinkedHashMap here instead of maintaining two data structures?

Reply:
I thought about it, but didn't like that LinkedHashMap doesn't provide a pop() mechanism like Queue. And I didn't really like the idea of iterating over all elements every time I wanted the last item.

Reply:
I was mistaken about LinkedHashMap ordering; going to swap one in